storage/localstore/localstore.go

```diff
@@ -336,6 +342,50 @@ func New(path string, baseKey []byte, o *Options) (db *DB, err error) {
 	if err != nil {
 		return nil, err
 	}
+
+	// Create a index structure for storing pinned chunks and their pin counts
+	db.pinIndex, err = db.shed.NewIndex("Hash->pinCounter", shed.IndexFuncs{
+		EncodeKey: func(fields shed.Item) (key []byte, err error) {
+			return fields.Address, nil
+		},
+		DecodeKey: func(key []byte) (e shed.Item, err error) {
+			e.Address = key
+			return e, nil
+		},
+		EncodeValue: func(fields shed.Item) (value []byte, err error) {
+			b := make([]byte, 8)
+			binary.BigEndian.PutUint64(b[:8], fields.PinCounter)
+			return b, nil
+		},
+		DecodeValue: func(keyItem shed.Item, value []byte) (e shed.Item, err error) {
+			e.PinCounter = binary.BigEndian.Uint64(value[:8])
+			return e, nil
+		},
+	})
+
+	// Create a index structure for storing the root hashes of the pinned files / manifests
+	db.pinFilesIndex, err = db.shed.NewIndex("Hash->IsRaw", shed.IndexFuncs{
```

Comment:
let's discuss the necessity of this index

Reply:
When someone issues a pin command like "swarm pin <filehash>" or "swarm up --pin <filename>", the root hash of what is pinned needs to be recorded somewhere. Because

1) When someone issues command "swarm pin ls" , we should show all the pinned root hashes and their info.  

2) When someone issues command "swarm pin remove <pinned hash>", we should check if this file is pinned.. only then unpin it... otherwise there is a possibility of chunks getting -ve pin counter values. Also i explained the IsRAW need above..

Reply:
I think it is the same as tags, should be handled the same way. @acud?

Reply:
We talked about this earlier on the call that If we handle this like Tags.. there needs to be an extra index which is storage expensive. Instead i use the root hash and store  the isRaw to walk the manifest..  I use this as a KV storage to store the list of file pinned (to be shown in ListPinnedFiles() command).

Reply:
Moved this to state store.